src/2025h2/cargo-caching-improvements.md

```diff
@@ -0,0 +1,111 @@
+| Point of contact | @ranger-ross                                                                     |
+| Teams            | &lt;!-- TEAMS WITH ASKS --&gt;                                                   |
+| Task owners      | &lt;!-- TASK OWNERS --&gt;                                                       |
+| Status           | Proposed                                                                         |
+| Tracking issue   | *if this is a continuing goal, add the old tracking issue, else leave blank*     |
+| Zulip channel    | N/A (an existing stream can be re-used or new streams can be created on request) |
+
+## Summary
+
+Rework the Cargo build directory layout to enable a shared build cache across multiple workspaces.
+
+
+## Motivation
+
+The primary goal of this effort is to improve build times by reusing builds across projects.
+
+Secondary goals are
+
+* Reduce disk usage of build artifacts
+* More precise cross-job caching in CI
+
+### The status quo
+
+When Cargo performs a build, it will build the package you requested and all
+dependencies individually, linking them in the end.
+These build results (intermediate build artifacts) and the linked result (final
+build artifact) are stored in the target-dir, which is per-workspace by
+default.
+
+Ways cargo will try to reuse builds today:
+- On a subsequent build, Cargo tries to reuse these build results by
+  "fingerprinting" the inputs to the prior build and checking if that
+  fingerprint has changed.
+- When dependencies are shared by host (`build.rs`, proc-macros) and
+  platform-target and the platform-target is the host, Cargo will attempt to
+  share host/target builds
+
+Some users try to get extra cache reuse by assigning all workspaces to use the same target-dir.
+- Cross-project conflicts occur because this shares both intermediate (generally unique) and final build artifacts (might not be unique)
+- `cargo clean` will clear the entire cache for every project
+- Rebuild churn from build inputs, like `RUSTFLAGS`, that cause a rebuild but aren't hashed into the file path
+
+In CI, users generally have to declare what directory should be cached between jobs.
+This directory will be compressed and uploaded at the end of the job.
+If the next job's cache key matches, the tarball will be downloaded and decompressed.
+If too much is cached, the time for managing the cache can dwarf the benefits of the cache.
+Some third-party projects exist to help manage cache size.
+
+### The next 6 months
+
+* Stabilize Cargo `build-dir` which allows users to move intermediate build artifacts out of `target` (rust-lang/cargo#14125)
+* Rework the build directory layout so that we have more easily cacheable "units" (rust-lang/cargo#15010)
+* Work towards a user wide build cache that can be reused across projects/workspaces. (rust-lang/cargo#5931)
+* Explore possibilities of extending this build cache to support remote storage enabling continuous integration use cases.
+
```

Comment:
Something we discussed today in the cargo team meeting is that this seems a little ambitious for a 6-month goal. I'm not sure what changes would be appropriate, but maybe @epage has some ideas.